// @airgap/beacon-transport-walletconnect/dist/cjs/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidNetworkOrAccount = exports.ActiveAccountUnspecified = exports.ActiveNetworkUnspecified = exports.MissingRequiredScope = exports.InvalidSession = exports.NotConnected = exports.InvalidReceivedSessionNamespace = void 0;
/**
 *  @category Error
 *  @description Error that indicates the wallet returned an invalid namespace
 */
class InvalidReceivedSessionNamespace extends Error {
    constructor(messageWc, codeWc, type, data) {
        super();
        this.messageWc = messageWc;
        this.codeWc = codeWc;
        this.data = data;
        this.name = 'InvalidReceivedSessionNamespace';
        const baseMessage = `${codeWc}: ${messageWc}.`;
        this.message = data
            ? type === 'incomplete'
                ? ` ${baseMessage} "${data}" is missing in the session namespace.`
                : ` ${baseMessage} "${data}" is invalid.`
            : baseMessage;
    }
}
exports.InvalidReceivedSessionNamespace = InvalidReceivedSessionNamespace;
/**
 *  @category Error
 *  @description Error that indicates there is no active session
 */
class NotConnected extends Error {
    constructor() {
        super('Not connected, no active session');
        this.name = 'NotConnected';
    }
}
exports.NotConnected = NotConnected;
/**
 *  @category Error
 *  @description Error that indicates the session is invalid
 */
class InvalidSession extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidSession';
    }
}
exports.InvalidSession = InvalidSession;
/**
 *  @category Error
 *  @description Error that indicates missing required permission scopes
 */
class MissingRequiredScope extends Error {
    constructor(requiredScopes) {
        super(`Required permission scope were not granted for "${requiredScopes}"`);
        this.requiredScopes = requiredScopes;
        this.name = 'MissingRequiredScope';
    }
}
exports.MissingRequiredScope = MissingRequiredScope;
/**
 *  @category Error
 *  @description Error that indicates the active network is not specified
 */
class ActiveNetworkUnspecified extends Error {
    constructor() {
        super('Please specify the active network using the "setActiveNetwork" method.');
        this.name = 'ActiveNetworkUnspecified';
    }
}
exports.ActiveNetworkUnspecified = ActiveNetworkUnspecified;
/**
 *  @category Error
 *  @description Error that indicates the active account is not specified
 */
class ActiveAccountUnspecified extends Error {
    constructor() {
        super('Please specify the active account using the "setActiveAccount" method.');
        this.name = 'ActiveAccountUnspecified';
    }
}
exports.ActiveAccountUnspecified = ActiveAccountUnspecified;
/**
 *  @category Error
 *  @description Error that indicates the combination pkh-network is not part of the active session
 */
class InvalidNetworkOrAccount extends Error {
    constructor(network, pkh) {
        super(`No permission. The combination "${network}" and "${pkh}" is not part of the active session.`);
        this.network = network;
        this.pkh = pkh;
        this.name = 'InvalidNetworkOrAccount';
    }
}
exports.InvalidNetworkOrAccount = InvalidNetworkOrAccount;
