// @taquito/taquito/dist/lib/operations/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasMetadataWithInternalOperationResult = exports.hasMetadataWithResult = exports.hasMetadata = exports.isSourceOp = exports.isOpRequireReveal = exports.isOpWithFee = exports.isOpWithGasBuffer = exports.isKind = exports.findWithKind = exports.attachKind = exports.OpKind = void 0;
const rpc_1 = require("@taquito/rpc");
var rpc_2 = require("@taquito/rpc");
Object.defineProperty(exports, "OpKind", { enumerable: true, get: function () { return rpc_2.OpKind; } });
const attachKind = (op, kind) => {
    return Object.assign(Object.assign({}, op), { kind });
};
exports.attachKind = attachKind;
const findWithKind = (arr, kind) => {
    if (Array.isArray(arr)) {
        const found = arr.find((op) => op.kind === kind);
        if (found && (0, exports.isKind)(found, kind)) {
            return found;
        }
    }
};
exports.findWithKind = findWithKind;
const isKind = (op, kind) => {
    return op.kind === kind;
};
exports.isKind = isKind;
const isOpWithGasBuffer = (op) => {
    if (op.kind === rpc_1.OpKind.TRANSACTION && op.parameters) {
        return true;
    }
    else {
        return ([
            'origination',
            'register_global_constant',
            'transfer_ticket',
            'update_consensus_key',
            'smart_rollup_add_messages',
            'smart_rollup_originate',
        ].indexOf(op.kind) !== -1);
    }
};
exports.isOpWithGasBuffer = isOpWithGasBuffer;
const isOpWithFee = (op) => {
    return ([
        'transaction',
        'delegation',
        'origination',
        'reveal',
        'register_global_constant',
        'increase_paid_storage',
        'transfer_ticket',
        'update_consensus_key',
        'smart_rollup_add_messages',
        'smart_rollup_originate',
        'smart_rollup_execute_outbox_message',
    ].indexOf(op.kind) !== -1);
};
exports.isOpWithFee = isOpWithFee;
const isOpRequireReveal = (op) => {
    return ([
        'transaction',
        'delegation',
        'origination',
        'register_global_constant',
        'increase_paid_storage',
        'transfer_ticket',
        'update_consensus_key',
        'smart_rollup_add_messages',
        'smart_rollup_originate',
        'smart_rollup_execute_outbox_message',
    ].indexOf(op.kind) !== -1);
};
exports.isOpRequireReveal = isOpRequireReveal;
const isSourceOp = (op) => {
    return ['transaction', 'delegation', 'origination', 'reveal', 'ballot'].indexOf(op.kind) !== -1;
};
exports.isSourceOp = isSourceOp;
const hasMetadata = (op) => {
    return 'metadata' in op;
};
exports.hasMetadata = hasMetadata;
const hasMetadataWithResult = (op) => {
    return (0, exports.hasMetadata)(op) && 'operation_result' in op.metadata;
};
exports.hasMetadataWithResult = hasMetadataWithResult;
const hasMetadataWithInternalOperationResult = (op) => {
    return (0, exports.hasMetadata)(op) && 'internal_operation_results' in op.metadata;
};
exports.hasMetadataWithInternalOperationResult = hasMetadataWithInternalOperationResult;
