// @airgap/beacon-types/dist/cjs/types/PostMessagePairingResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendedPostMessagePairingResponse = exports.PostMessagePairingResponse = void 0;
/**
 * @internalapi
 */
class PostMessagePairingResponse {
    constructor(id, name, publicKey, version, icon, appUrl) {
        this.type = 'postmessage-pairing-response';
        this.id = id;
        this.name = name;
        this.icon = icon;
        this.appUrl = appUrl;
        this.publicKey = publicKey;
        this.version = version;
    }
}
exports.PostMessagePairingResponse = PostMessagePairingResponse;
/**
 * @internalapi
 */
class ExtendedPostMessagePairingResponse extends PostMessagePairingResponse {
    constructor(id, name, publicKey, version, senderId, extensionId, icon, appUrl) {
        super(id, name, publicKey, version, icon, appUrl);
        this.senderId = senderId;
        this.extensionId = extensionId;
    }
}
exports.ExtendedPostMessagePairingResponse = ExtendedPostMessagePairingResponse;
// TODO: Rename to "WalletPeerInfo"?
