// @airgap/beacon-types/dist/cjs/types/WalletConnectPairingResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendedWalletConnectPairingResponse = exports.WalletConnectPairingResponse = void 0;
/**
 * @internalapi
 */
class WalletConnectPairingResponse {
    constructor(id, name, publicKey, version, icon, appUrl) {
        this.type = 'walletconnect-pairing-response';
        this.id = id;
        this.name = name;
        this.icon = icon;
        this.appUrl = appUrl;
        this.publicKey = publicKey;
        this.version = version;
    }
}
exports.WalletConnectPairingResponse = WalletConnectPairingResponse;
/**
 * @internalapi
 */
class ExtendedWalletConnectPairingResponse extends WalletConnectPairingResponse {
    constructor(id, name, publicKey, version, senderId, extensionId, icon, appUrl) {
        super(id, name, publicKey, version, icon, appUrl);
        this.senderId = senderId;
        this.extensionId = extensionId;
    }
}
exports.ExtendedWalletConnectPairingResponse = ExtendedWalletConnectPairingResponse;
// TODO: Rename to "WalletPeerInfo"?
