// @airgap/beacon-types/dist/esm/types/tezos/operations/SmartRollupRefute.js
export var SmartRollupRefutationOptions;
(function (SmartRollupRefutationOptions) {
    SmartRollupRefutationOptions["START"] = "start";
    SmartRollupRefutationOptions["MOVE"] = "move";
})(SmartRollupRefutationOptions || (SmartRollupRefutationOptions = {}));
export var SmartRollupRefuteGameStatusOptions;
(function (SmartRollupRefuteGameStatusOptions) {
    SmartRollupRefuteGameStatusOptions["ONGOING"] = "ongoing";
    SmartRollupRefuteGameStatusOptions["ENDED"] = "ended";
})(SmartRollupRefuteGameStatusOptions || (SmartRollupRefuteGameStatusOptions = {}));
export var SmartRollupInputProofKind;
(function (SmartRollupInputProofKind) {
    SmartRollupInputProofKind["INBOX_PROOF"] = "inbox_proof";
    SmartRollupInputProofKind["REVEAL_PROOF"] = "reveal_proof";
    SmartRollupInputProofKind["FIRST_INPUT"] = "first_input";
})(SmartRollupInputProofKind || (SmartRollupInputProofKind = {}));
export var SmartRollupRefuteGameEndedPlayerOutcomes;
(function (SmartRollupRefuteGameEndedPlayerOutcomes) {
    SmartRollupRefuteGameEndedPlayerOutcomes["LOSER"] = "loser";
    SmartRollupRefuteGameEndedPlayerOutcomes["DRAW"] = "draw";
})(SmartRollupRefuteGameEndedPlayerOutcomes || (SmartRollupRefuteGameEndedPlayerOutcomes = {}));
export var SmartRollupRefuteGameEndedReason;
(function (SmartRollupRefuteGameEndedReason) {
    SmartRollupRefuteGameEndedReason["CONFLICT_RESOLVED"] = "conflict_resolved";
    SmartRollupRefuteGameEndedReason["TIMEOUT"] = "timeout";
})(SmartRollupRefuteGameEndedReason || (SmartRollupRefuteGameEndedReason = {}));
export var SmartRollupRefuteRevealProofKind;
(function (SmartRollupRefuteRevealProofKind) {
    SmartRollupRefuteRevealProofKind["RAW_DATA_PROOF"] = "raw_data_proof";
    SmartRollupRefuteRevealProofKind["METADATA_PROOF"] = "metadata_proof";
    SmartRollupRefuteRevealProofKind["DAL_PAGE_PROOF"] = "dal_page_proof";
})(SmartRollupRefuteRevealProofKind || (SmartRollupRefuteRevealProofKind = {}));
