// @airgap/beacon-core/dist/cjs/managers/AccountManager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountManager = void 0;
const beacon_types_1 = require("@airgap/beacon-types");
const StorageManager_1 = require("./StorageManager");
const PermissionValidator_1 = require("./PermissionValidator");
/**
 * @internalapi
 *
 * The AccountManager provides CRUD functionality for account entities and persists them to the provided storage.
 */
class AccountManager {
    constructor(storage) {
        this.storageManager = new StorageManager_1.StorageManager(storage, beacon_types_1.StorageKey.ACCOUNTS);
    }
    getAccounts() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            return (_a = (yield this.storageManager.getAll())) !== null && _a !== void 0 ? _a : [];
        });
    }
    getAccount(accountIdentifier) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageManager.getOne((account) => account.accountIdentifier === accountIdentifier);
        });
    }
    addAccount(accountInfo) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageManager.addOne(accountInfo, (account) => account.accountIdentifier === accountInfo.accountIdentifier);
        });
    }
    updateAccount(accountIdentifier, accountInfo) {
        return __awaiter(this, void 0, void 0, function* () {
            const account = yield this.getAccount(accountIdentifier);
            if (!account)
                return undefined;
            const newAccount = Object.assign(Object.assign({}, account), accountInfo);
            yield this.storageManager.addOne(newAccount, (account) => account.accountIdentifier === accountIdentifier, true);
            return newAccount;
        });
    }
    removeAccount(accountIdentifier) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageManager.remove((account) => account.accountIdentifier === accountIdentifier);
        });
    }
    removeAccounts(accountIdentifiers) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageManager.remove((account) => accountIdentifiers.includes(account.accountIdentifier));
        });
    }
    removeAllAccounts() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.storageManager.removeAll();
        });
    }
    hasPermission(message) {
        return __awaiter(this, void 0, void 0, function* () {
            return PermissionValidator_1.PermissionValidator.hasPermission(message, this.getAccount.bind(this), this.getAccounts.bind(this));
        });
    }
}
exports.AccountManager = AccountManager;
