// @airgap/beacon-types/dist/cjs/types/tezos/OperationTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezosOperationType = void 0;
/**
 * @publicapi
 * @category Tezos
 */
var TezosOperationType;
(function (TezosOperationType) {
    TezosOperationType["ORIGINATION"] = "origination";
    TezosOperationType["DELEGATION"] = "delegation";
    TezosOperationType["REVEAL"] = "reveal";
    TezosOperationType["TRANSACTION"] = "transaction";
    TezosOperationType["ACTIVATE_ACCOUNT"] = "activate_account";
    TezosOperationType["ENDORSEMENT"] = "endorsement";
    TezosOperationType["SEED_NONCE_REVELATION"] = "seed_nonce_revelation";
    TezosOperationType["DOUBLE_ENDORSEMENT_EVIDENCE"] = "double_endorsement_evidence";
    TezosOperationType["DOUBLE_BAKING_EVIDENCE"] = "double_baking_evidence";
    TezosOperationType["PROPOSALS"] = "proposals";
    TezosOperationType["BALLOT"] = "ballot";
    TezosOperationType["ATTESTATION"] = "attestation";
    TezosOperationType["PREATTESTATION"] = "preattestation";
    TezosOperationType["PREENDORSEMENT"] = "preendorsement";
    TezosOperationType["SET_DEPOSITS_LIMIT"] = "set_deposits_limit";
    TezosOperationType["DOUBLE_PREATTESTATION_EVIDENCE"] = "double_preattestation_evidence";
    TezosOperationType["DOUBLE_PREENDORSEMENT_EVIDENCE"] = "double_preendorsement_evidence";
    TezosOperationType["ATTESTATION_WITH_SLOT"] = "attestation_with_slot";
    TezosOperationType["ENDORSEMENT_WITH_SLOT"] = "endorsement_with_slot";
    TezosOperationType["DOUBLE_ATTESTATION_EVIDENCE"] = "double_attestation_evidence";
    TezosOperationType["FAILING_NOOP"] = "failing_noop";
    TezosOperationType["REGISTER_GLOBAL_CONSTANT"] = "register_global_constant";
    TezosOperationType["TRANSFER_TICKET"] = "transfer_ticket";
    TezosOperationType["INCREASE_PAID_STORAGE"] = "increase_paid_storage";
    TezosOperationType["UPDATE_CONSENSUS_KEY"] = "update_consensus_key";
    TezosOperationType["DRAIN_DELEGATE"] = "drain_delegate";
    TezosOperationType["VDF_REVELATION"] = "vdf_revelation";
    TezosOperationType["EVENT"] = "event";
    TezosOperationType["TICKET_UPDATES"] = "ticket_updates";
    TezosOperationType["SMART_ROLLUP_ORIGINATE"] = "smart_rollup_originate";
    TezosOperationType["SMART_ROLLUP_ADD_MESSAGES"] = "smart_rollup_add_messages";
    TezosOperationType["SMART_ROLLUP_EXECUTE_OUTBOX_MESSAGE"] = "smart_rollup_execute_outbox_message";
    TezosOperationType["SMART_ROLLUP_PUBLISH"] = "smart_rollup_publish";
    TezosOperationType["SMART_ROLLUP_CEMENT"] = "smart_rollup_cement";
    TezosOperationType["SMART_ROLLUP_RECOVER_BOND"] = "smart_rollup_recover_bond";
    TezosOperationType["SMART_ROLLUP_REFUTE"] = "smart_rollup_refute";
    TezosOperationType["SMART_ROLLUP_TIMEOUT"] = "smart_rollup_timeout";
    TezosOperationType["DAL_PUBLISH_COMMITMENT"] = "dal_publish_commitment";
})(TezosOperationType || (exports.TezosOperationType = TezosOperationType = {}));
