// @airgap/beacon-core/dist/cjs/storage/StorageValidator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageValidator = void 0;
const beacon_types_1 = require("@airgap/beacon-types");
class StorageValidator {
    constructor(storage) {
        this.storage = storage;
    }
    validateNumber(param) {
        return typeof param === 'number' && !isNaN(param);
    }
    validateText(param) {
        return typeof param === 'string';
    }
    validateBoolean(param) {
        return typeof param === 'boolean';
    }
    validateArray(param) {
        return Array.isArray(param);
    }
    objHasProperty(obj, path) {
        if (!obj)
            return false; // Return false if the object is null or undefined
        const properties = path.split('.'); // Split the path into individual properties
        let current = obj;
        for (const property of properties) {
            // If the property doesn't exist, return false
            if (!current.hasOwnProperty(property)) {
                return false;
            }
            // Move to the next level in the path
            current = current[property];
        }
        return true;
    }
    innerValidate(value, type, prop) {
        if (!value) {
            return true;
        }
        switch (type) {
            case 'num':
                return this.validateNumber(value);
            case 'str':
                return this.validateText(value);
            case 'bol':
                return this.validateBoolean(value);
            case 'obj':
                return this.objHasProperty(value, prop);
            case 'arr':
                return this.validateArray(value);
            default:
                return false;
        }
    }
    validate() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.BEACON_SDK_VERSION), 'str')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.MATRIX_SELECTED_NODE), 'str')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.MULTI_NODE_SETUP_DONE), 'bol')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.TRANSPORT_P2P_PEERS_DAPP), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.TRANSPORT_P2P_PEERS_WALLET), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.TRANSPORT_POSTMESSAGE_PEERS_DAPP), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.TRANSPORT_POSTMESSAGE_PEERS_WALLET), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.TRANSPORT_WALLETCONNECT_PEERS_DAPP), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.ACCOUNTS), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.APP_METADATA_LIST), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.PERMISSION_LIST), 'arr')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.ACTIVE_ACCOUNT), 'str')) {
                return false;
            }
            if (!this.innerValidate(yield this.storage.get(beacon_types_1.StorageKey.LAST_SELECTED_WALLET), 'obj', 'key')) {
                return false;
            }
            return true;
        });
    }
}
exports.StorageValidator = StorageValidator;
