// @taquito/rpc/dist/lib/rpc-client-interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCMethodName = exports.defaultRPCOptions = exports.defaultChain = void 0;
exports.defaultChain = 'main';
exports.defaultRPCOptions = { block: 'head' };
var RPCMethodName;
(function (RPCMethodName) {
    RPCMethodName["GET_BAKING_RIGHTS"] = "getBakingRights";
    RPCMethodName["GET_BALLOTS"] = "getBallots";
    RPCMethodName["GET_BALLOT_LIST"] = "getBallotList";
    RPCMethodName["GET_BIG_MAP_KEY"] = "getBigMapKey";
    RPCMethodName["GET_BIG_MAP_EXPR"] = "getBigMapExpr";
    RPCMethodName["GET_BLOCK_HASH"] = "getBlockHash";
    RPCMethodName["GET_BLOCK"] = "getBlock";
    RPCMethodName["GET_BLOCK_HEADER"] = "getBlockHeader";
    RPCMethodName["GET_BLOCK_METADATA"] = "getBlockMetadata";
    RPCMethodName["GET_BALANCE"] = "getBalance";
    RPCMethodName["GET_FULL_BALANCE"] = "getFullBalance";
    RPCMethodName["GET_STAKED_BALANCE"] = "getStakedBalance";
    RPCMethodName["GET_UNSTAKED_FINALIZABLE_BALANCE"] = "getUnstakedFinalizableBalance";
    RPCMethodName["GET_UNSTAKED_FROZEN_BALANCE"] = "getUnstakedFrozenBalance";
    RPCMethodName["GET_UNSTAKE_REQUESTS"] = "getUnstakeRequests";
    RPCMethodName["GET_CHAIN_ID"] = "getChainId";
    RPCMethodName["GET_CONSTANTS"] = "getConstants";
    RPCMethodName["GET_CONTRACT"] = "getContract";
    RPCMethodName["GET_CURRENT_PERIOD"] = "getCurrentPeriod";
    RPCMethodName["GET_CURRENT_PROPOSAL"] = "getCurrentProposal";
    RPCMethodName["GET_CURRENT_QUORUM"] = "getCurrentQuorum";
    RPCMethodName["GET_DELEGATE"] = "getDelegate";
    RPCMethodName["GET_ALL_DELEGATES"] = "getAllDelegates";
    RPCMethodName["GET_DELEGATES"] = "getDelegates";
    RPCMethodName["GET_VOTING_INFO"] = "getVotingInfo";
    RPCMethodName["GET_ATTESTATION_RIGHTS"] = "getAttestationRights";
    RPCMethodName["GET_ENTRYPOINTS"] = "getEntrypoints";
    RPCMethodName["GET_LIVE_BLOCKS"] = "getLiveBlocks";
    RPCMethodName["GET_MANAGER_KEY"] = "getManagerKey";
    RPCMethodName["GET_NORMALIZED_SCRIPT"] = "getNormalizedScript";
    RPCMethodName["GET_PROPOSALS"] = "getProposals";
    RPCMethodName["GET_PROTOCOLS"] = "getProtocols";
    RPCMethodName["GET_SAPLING_DIFF_BY_CONTRACT"] = "getSaplingDiffByContract";
    RPCMethodName["GET_SAPLING_DIFF_BY_ID"] = "getSaplingDiffById";
    RPCMethodName["GET_SCRIPT"] = "getScript";
    RPCMethodName["GET_STORAGE"] = "getStorage";
    RPCMethodName["GET_SUCCESSOR_PERIOD"] = "getSuccessorPeriod";
    RPCMethodName["GET_VOTES_LISTINGS"] = "getVotesListings";
    RPCMethodName["PACK_DATA"] = "packData";
    RPCMethodName["GET_STORAGE_USED_SPACE"] = "getStorageUsedSpace";
    RPCMethodName["GET_STORAGE_PAID_SPACE"] = "getStoragePaidSpace";
    RPCMethodName["GET_TICKET_BALANCE"] = "getTicketBalance";
    RPCMethodName["GET_ALL_TICKET_BALANCES"] = "getAllTicketBalances";
    RPCMethodName["GET_ADAPTIVE_ISSUANCE_LAUNCH_CYCLE"] = "getAdaptiveIssuanceLaunchCycle";
    RPCMethodName["GET_PENDING_OPERATIONS"] = "getPendingOperations";
})(RPCMethodName || (exports.RPCMethodName = RPCMethodName = {}));
