// @airgap/beacon-types/dist/cjs/types/tezos/operations/SmartRollupRefute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartRollupRefuteRevealProofKind = exports.SmartRollupRefuteGameEndedReason = exports.SmartRollupRefuteGameEndedPlayerOutcomes = exports.SmartRollupInputProofKind = exports.SmartRollupRefuteGameStatusOptions = exports.SmartRollupRefutationOptions = void 0;
var SmartRollupRefutationOptions;
(function (SmartRollupRefutationOptions) {
    SmartRollupRefutationOptions["START"] = "start";
    SmartRollupRefutationOptions["MOVE"] = "move";
})(SmartRollupRefutationOptions || (exports.SmartRollupRefutationOptions = SmartRollupRefutationOptions = {}));
var SmartRollupRefuteGameStatusOptions;
(function (SmartRollupRefuteGameStatusOptions) {
    SmartRollupRefuteGameStatusOptions["ONGOING"] = "ongoing";
    SmartRollupRefuteGameStatusOptions["ENDED"] = "ended";
})(SmartRollupRefuteGameStatusOptions || (exports.SmartRollupRefuteGameStatusOptions = SmartRollupRefuteGameStatusOptions = {}));
var SmartRollupInputProofKind;
(function (SmartRollupInputProofKind) {
    SmartRollupInputProofKind["INBOX_PROOF"] = "inbox_proof";
    SmartRollupInputProofKind["REVEAL_PROOF"] = "reveal_proof";
    SmartRollupInputProofKind["FIRST_INPUT"] = "first_input";
})(SmartRollupInputProofKind || (exports.SmartRollupInputProofKind = SmartRollupInputProofKind = {}));
var SmartRollupRefuteGameEndedPlayerOutcomes;
(function (SmartRollupRefuteGameEndedPlayerOutcomes) {
    SmartRollupRefuteGameEndedPlayerOutcomes["LOSER"] = "loser";
    SmartRollupRefuteGameEndedPlayerOutcomes["DRAW"] = "draw";
})(SmartRollupRefuteGameEndedPlayerOutcomes || (exports.SmartRollupRefuteGameEndedPlayerOutcomes = SmartRollupRefuteGameEndedPlayerOutcomes = {}));
var SmartRollupRefuteGameEndedReason;
(function (SmartRollupRefuteGameEndedReason) {
    SmartRollupRefuteGameEndedReason["CONFLICT_RESOLVED"] = "conflict_resolved";
    SmartRollupRefuteGameEndedReason["TIMEOUT"] = "timeout";
})(SmartRollupRefuteGameEndedReason || (exports.SmartRollupRefuteGameEndedReason = SmartRollupRefuteGameEndedReason = {}));
var SmartRollupRefuteRevealProofKind;
(function (SmartRollupRefuteRevealProofKind) {
    SmartRollupRefuteRevealProofKind["RAW_DATA_PROOF"] = "raw_data_proof";
    SmartRollupRefuteRevealProofKind["METADATA_PROOF"] = "metadata_proof";
    SmartRollupRefuteRevealProofKind["DAL_PAGE_PROOF"] = "dal_page_proof";
})(SmartRollupRefuteRevealProofKind || (exports.SmartRollupRefuteRevealProofKind = SmartRollupRefuteRevealProofKind = {}));
