// @taquito/taquito/dist/lib/wallet/legacy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyWalletProvider = void 0;
const types_1 = require("../operations/types");
class LegacyWalletProvider {
    constructor(context) {
        this.context = context;
    }
    getPKH() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.context.signer.publicKeyHash();
        });
    }
    getPK() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.context.signer.publicKey();
        });
    }
    mapTransferParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.TRANSACTION);
        });
    }
    mapStakeParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.TRANSACTION);
        });
    }
    mapUnstakeParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.TRANSACTION);
        });
    }
    mapFinalizeUnstakeParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.TRANSACTION);
        });
    }
    mapOriginateParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.ORIGINATION);
        });
    }
    mapDelegateParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.DELEGATION);
        });
    }
    mapIncreasePaidStorageWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.INCREASE_PAID_STORAGE);
        });
    }
    mapTransferTicketParamsToWalletParams(params) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, types_1.attachKind)(yield params(), types_1.OpKind.TRANSFER_TICKET);
        });
    }
    sendOperations(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const op = yield this.context.batch.batch(params).send();
            return op.hash;
        });
    }
    sign(bytes, watermark) {
        return __awaiter(this, void 0, void 0, function* () {
            const { prefixSig } = yield this.context.signer.sign(bytes, watermark);
            return prefixSig;
        });
    }
}
exports.LegacyWalletProvider = LegacyWalletProvider;
