// @airgap/beacon-core/dist/cjs/MockWindow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearMockWindowState = exports.windowRef = void 0;
const cbs = [(_) => undefined];
/**
 * A mock for postmessage if run in node.js environment
 */
let windowRef = {
    postMessage: (message, _target) => {
        console.log('GOT MOCK POST MESSAGE', message);
        cbs.forEach((callbackElement) => {
            callbackElement({ data: message });
        });
    },
    addEventListener: (_name, eventCallback) => {
        cbs.push(eventCallback);
    },
    removeEventListener: (_name, eventCallback) => {
        cbs.splice(cbs.indexOf((element) => element === eventCallback), 1);
    },
    location: {
        origin: '*'
    }
};
exports.windowRef = windowRef;
try {
    if (typeof window !== 'undefined') {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        exports.windowRef = windowRef = window;
    }
}
catch (windowError) {
    console.log(`not defined: ${windowError}`);
}
const clearMockWindowState = () => {
    cbs.length = 0;
};
exports.clearMockWindowState = clearMockWindowState;
