// @airgap/beacon-core/dist/cjs/errors/BeaconError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeaconError = void 0;
const assert_never_1 = require("../utils/assert-never");
const __1 = require("..");
const beacon_types_1 = require("@airgap/beacon-types");
/**
 * @category Error
 */
class BeaconError {
    get fullDescription() {
        return { description: this.description };
    }
    constructor(errorType, message) {
        this.name = 'BeaconError';
        this.title = 'Error'; // Visible in the UI
        this.message = `[${errorType}]:${message}`;
        this.description = message;
    }
    static getError(errorType, errorData) {
        switch (errorType) {
            case beacon_types_1.BeaconErrorType.BROADCAST_ERROR:
                return new __1.BroadcastBeaconError();
            case beacon_types_1.BeaconErrorType.NETWORK_NOT_SUPPORTED:
                return new __1.NetworkNotSupportedBeaconError();
            case beacon_types_1.BeaconErrorType.NO_ADDRESS_ERROR:
                return new __1.NoAddressBeaconError();
            case beacon_types_1.BeaconErrorType.NO_PRIVATE_KEY_FOUND_ERROR:
                return new __1.NoPrivateKeyBeaconError();
            case beacon_types_1.BeaconErrorType.NOT_GRANTED_ERROR:
                return new __1.NotGrantedBeaconError();
            case beacon_types_1.BeaconErrorType.PARAMETERS_INVALID_ERROR:
                return new __1.ParametersInvalidBeaconError();
            case beacon_types_1.BeaconErrorType.TOO_MANY_OPERATIONS:
                return new __1.TooManyOperationsBeaconError();
            case beacon_types_1.BeaconErrorType.TRANSACTION_INVALID_ERROR:
                return new __1.TransactionInvalidBeaconError(errorData);
            case beacon_types_1.BeaconErrorType.SIGNATURE_TYPE_NOT_SUPPORTED:
                return new __1.SignatureTypeNotSupportedBeaconError();
            // case BeaconErrorType.ENCRYPTION_TYPE_NOT_SUPPORTED:
            //   return new EncryptionTypeNotSupportedBeaconError()
            case beacon_types_1.BeaconErrorType.ABORTED_ERROR:
                return new __1.AbortedBeaconError();
            case beacon_types_1.BeaconErrorType.UNKNOWN_ERROR:
                return new __1.UnknownBeaconError();
            default:
                (0, assert_never_1.assertNever)(errorType);
        }
    }
}
exports.BeaconError = BeaconError;
